django/db/migrations/state.py

```diff
@@ -187,38 +211,54 @@ def add_field(self, app_label, model_name, name, field, preserve_default):
             field.default = NOT_PROVIDED
         else:
             field = field
-        self.models[app_label, model_name].fields[name] = field
+        model_key = app_label, model_name
+        self.models[model_key].fields[name] = field
+        if self._relations is not None:
+            self.resolve_model_field_relations(model_key, name, field)
         # Delay rendering of relationships if it's not a relational field.
         delay = not field.is_relation
-        self.reload_model(app_label, model_name, delay=delay)
+        self.reload_model(*model_key, delay=delay)
 
     def remove_field(self, app_label, model_name, name):
-        model_state = self.models[app_label, model_name]
+        model_key = app_label, model_name
+        model_state = self.models[model_key]
         old_field = model_state.fields.pop(name)
+        if self._relations is not None:
+            self.resolve_model_field_relations(model_key, name, old_field)
         # Delay rendering of relationships if it's not a relational field.
         delay = not old_field.is_relation
-        self.reload_model(app_label, model_name, delay=delay)
+        self.reload_model(*model_key, delay=delay)
 
     def alter_field(self, app_label, model_name, name, field, preserve_default):
         if not preserve_default:
             field = field.clone()
             field.default = NOT_PROVIDED
         else:
             field = field
-        model_state = self.models[app_label, model_name]
-        model_state.fields[name] = field
+        model_key = app_label, model_name
+        model_state = self.models[model_key]
```

Comment:
It looks like you can actually do `fields = self.models[model_key].fields` (similar to further down) since you're only using `model_state.fields` below.